modules/transformer-kinesis/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/transformer/kinesis/sources/Kinesis.scala

```diff
@@ -56,4 +57,21 @@ object Kinesis {
       case InitPosition.TrimHorizon       => InitialPositionInStream.TRIM_HORIZON.asLeft
       case InitPosition.AtTimestamp(date) => Date.from(date).asRight
     }
+
+  def checkpointer[F[_]: Sync](record: CommittableRecord): KinesisCheckpointer[F] =
+    KinesisCheckpointer[F](Map(record.shardId -> record.checkpoint))
+
+
+  case class KinesisCheckpointer[F[_]](byShard: Map[String, F[Unit]])
+
+  object KinesisCheckpointer {
+    implicit def kinesisCheckpointer[F[_]: Applicative]: Checkpointer[F, KinesisCheckpointer[F]] = new Checkpointer[F, KinesisCheckpointer[F]] {
+      def checkpoint(c: KinesisCheckpointer[F]): F[Unit] = c.byShard.values.toList.sequence_
+      def combine(older: KinesisCheckpointer[F], newer: KinesisCheckpointer[F]): KinesisCheckpointer[F] =
+        KinesisCheckpointer[F](byShard = older.byShard ++ newer.byShard) // order is important!
```

Comment:
I do not need to preserve order of the keys.  I only need to make it so that right-hand values get selected over left-hand values.

For example, imagine if we have shards called "A", "B" and "C".  And records in the shards are called like "A100" or "B100".  Then if older and newer like this this:

```
older = {
  "shard-A": how-to-checkpoint-record-A100
  "shard-B": how-to-checkpoint-record-B100
}

newer = {
  "shard-A": how-to-checkpoint-record-A200
  "shard-C": how-to-checkpoint-record-C100
}
```

then I need `combine` to yield this:

```
{
  "shard-A": how-to-checkpoint-record-A200
  "shard-B": how-to-checkpoint-record-B100
  "shard-C": how-to-checkpoint-record-C300
}
```

When I issue it checkpoints, it does not matter on the order.  I can either checkpoint [A200, B100, C300] or I can checkpoint [B100, C300, A200].  Kinesis shards are checkpointed independently from each other.  We need to checkpoint every shard with the most recent record from each shard.